File: desktop/macos/Desktop/Sources/Bluetooth/Transports/DeviceTransport.swift
import Combine
import CoreBluetooth
import Foundation

/// Transport layer state for device communication
/// Ported from: omi/app/lib/services/devices/transports/device_transport.dart
enum DeviceTransportState: String, Sendable {
    case disconnected
    case connecting
    case connected
    case disconnecting
}

/// Abstract transport layer protocol for device communication
/// Provides a unified interface for different communication protocols (BLE, etc.)
/// Ported from: omi/app/lib/services/devices/transports/device_transport.dart
protocol DeviceTransport: AnyObject {

    /// Unique identifier for the connected device
    var deviceId: String { get }

    /// Current connection state
    var state: DeviceTransportState { get }

    /// Publisher for connection state changes
    var connectionStatePublisher: AnyPublisher<DeviceTransportState, Never> { get }

    /// Connect to the device
    func connect() async throws

    /// Disconnect from the device
    func disconnect() async

    /// Check if the device is currently connected
    func isConnected() async -> Bool

    /// Ping the device to verify connection is alive
    func ping() async -> Bool

    /// Get a stream of data from a characteristic
    /// - Parameters:
    ///   - serviceUUID: The service UUID
    ///   - characteristicUUID: The characteristic UUID
    /// - Returns: An AsyncStream of data from the characteristic
    func getCharacteristicStream(
        serviceUUID: CBUUID,
        characteristicUUID: CBUUID
    ) -> AsyncThrowingStream<Data, Error>

    /// Read a characteristic value
    /// - Parameters:
    ///   - serviceUUID: The service UUID
    ///   - characteristicUUID: The characteristic UUID
    /// - Returns: The characteristic data, or empty data if read fails
    func readCharacteristic(
        serviceUUID: CBUUID,
        characteristicUUID: CBUUID
    ) async throws -> Data

    /// Write data to a characteristic
    /// - Parameters:
    ///   - data: The data to write
    ///   - serviceUUID: The service UUID
    ///   - characteristicUUID: The characteristic UUID
    ///   - withResponse: Whether to wait for a write response
    func writeCharacteristic(
        data: Data,
        serviceUUID: CBUUID,
        characteristicUUID: CBUUID,
        withResponse: Bool
    ) async throws

    /// Clean up resources
    func dispose() async
}

// MARK: - Default Implementations

extension DeviceTransport {

    /// Write with response by default
    func writeCharacteristic(
        data: Data,
        serviceUUID: CBUUID,
        characteristicUUID: CBUUID
    ) async throws {
        try await writeCharacteristic(
            data: data,
            serviceUUID: serviceUUID,
            characteristicUUID: characteristicUUID,
            withResponse: true
        )
    }
}

// MARK: - Transport Errors

enum DeviceTransportError: LocalizedError {
    case notConnected
    case connectionFailed(String)
    case serviceNotFound(CBUUID)
    case characteristicNotFound(CBUUID)
    case readFailed(String)
    case writeFailed(String)
    case timeout
    case disposed

    var errorDescription: String? {
        switch self {
        case .notConnected:
            return "Device is not connected"
        case .connectionFailed(let reason):
            return "Connection failed: \(reason)"
        case .serviceNotFound(let uuid):
            return "Service not found: \(uuid.uuidString)"
        case .characteristicNotFound(let uuid):
            return "Characteristic not found: \(uuid.uuidString)"
        case .readFailed(let reason):
            return "Read failed: \(reason)"
        case .writeFailed(let reason):
            return "Write failed: \(reason)"
        case .timeout:
            return "Operation timed out"
        case .disposed:
            return "Transport has been disposed"
        }
    }
}

File: desktop/windows/src/renderer/src/device/transport/deviceTransport.ts
/**
 * Transport abstraction for wearable BLE connections — Windows port of macOS
 * Transports/DeviceTransport.swift. A transport is single-use: one physical
 * connection attempt per instance, fenced by a coordinator-assigned session
 * generation.
 */

export type DeviceTransportState = 'disconnected' | 'connecting' | 'connected' | 'disconnecting'

export type DeviceTransportErrorKind =
  | 'notConnected'
  | 'connectionFailed'
  | 'serviceNotFound'
  | 'characteristicNotFound'
  | 'readFailed'
  | 'writeFailed'
  | 'timeout'
  | 'disposed'

export class DeviceTransportError extends Error {
  readonly kind: DeviceTransportErrorKind

  private constructor(kind: DeviceTransportErrorKind, message: string) {
    super(message)
    this.name = 'DeviceTransportError'
    this.kind = kind
  }

  static notConnected(): DeviceTransportError {
    return new DeviceTransportError('notConnected', 'Device is not connected')
  }

  static connectionFailed(detail: string): DeviceTransportError {
    return new DeviceTransportError('connectionFailed', `Connection failed: ${detail}`)
  }

  static serviceNotFound(uuid: string): DeviceTransportError {
    return new DeviceTransportError('serviceNotFound', `Service not found: ${uuid}`)
  }

  static characteristicNotFound(uuid: string): DeviceTransportError {
    return new DeviceTransportError('characteristicNotFound', `Characteristic not found: ${uuid}`)
  }

  static readFailed(detail: string): DeviceTransportError {
    return new DeviceTransportError('readFailed', `Read failed: ${detail}`)
  }

  static writeFailed(detail: string): DeviceTransportError {
    return new DeviceTransportError('writeFailed', `Write failed: ${detail}`)
  }

  static timeout(): DeviceTransportError {
    return new DeviceTransportError('timeout', 'Operation timed out')
  }

  static disposed(): DeviceTransportError {
    return new DeviceTransportError('disposed', 'Transport has been disposed')
  }
}

export interface CharacteristicStreamSubscriber {
  onData(data: Uint8Array): void
  /** Exactly once. Null on clean transport shutdown; an error on mid-session
   *  disconnects so stream consumers observe connection loss. */
  onFinish(error: Error | null): void
}

export interface CharacteristicStreamSubscription {
  cancel(): void
}

export interface DeviceTransport {
  readonly deviceId: string
  readonly sessionGeneration: number
  readonly state: DeviceTransportState
  onStateChange(listener: (state: DeviceTransportState) => void): () => void
  connect(): Promise<void>
  disconnect(): Promise<void>
  isConnected(): boolean
  /** Liveness probe: true when the physical link currently looks connected.
   *  There is no keepalive loop anywhere in this stack. */
  ping(): boolean
  subscribeCharacteristic(
    serviceUuid: string,
    characteristicUuid: string,
    subscriber: CharacteristicStreamSubscriber
  ): CharacteristicStreamSubscription
  readCharacteristic(serviceUuid: string, characteristicUuid: string): Promise<Uint8Array>
  writeCharacteristic(args: {
    serviceUuid: string
    characteristicUuid: string
    data: Uint8Array
    /** Defaults to true (write-with-response). */
    withResponse?: boolean
  }): Promise<void>
  dispose(): Promise<void>
}
